### PR TITLE
HWY-191: Add basic tests for LNC validity.

### PR DESCRIPTION
**Pull request by goral09**
https://casperlabs.atlassian.net/browse/HWY-191

### CODE REVIEW

**Comment by goral09** [comment]
@afck @xcthulhu I've ignored the tests that have false positives (they expect `None` and stub returns `None`) and left the ones that fail. The issue is that we won't be able to merge the PR with failing tests. Do you want me to create a feature branch and merge it there? This will allow us to work off the same base.

File: node/src/components/consensus/highway_core/state.rs [hunk 1]
```diff
@@ -584,8 +586,20 @@ impl<C: Context> State<C> {
...
-        // TODO: Validate against LNC.
-        Ok(())
+        match self.validate_lnc(wunit) {
+            None => Ok(()),
+            Some(vidx) => Err(UnitError::LncNaiveCitation(vidx)),
+        }
+    }
+
+    /// Validates whether `wvote` violates the LNC rule.
+    /// Returns index of the first equivocator that should have been endorsed but wasn't.
```
Comment (afck) [inline node/src/components/consensus/highway_core/state.rs:596]:
```suggestion
    /// Returns index of the first equivocator that was cited naively twice
```
…or something like that? It's not the equivocator who needs to be endorsed.

**Review by afck — APPROVED** [review]

**Comment by afck** [comment]
> won't be able to merge the PR with failing tests

But if they are `#[ignore]`d, we can merge it, can't we?

**Comment by goral09** [comment]
> > won't be able to merge the PR with failing tests
> 
> But if they are `#[ignore]`d, we can merge it, can't we?

Tests that pass with stubbed `validate_lnc` are ignored. Tests that fail are not ignored. Do you want me to ignore all tests?

**Comment by afck** [comment]
> Tests that pass with stubbed validate_lnc are ignored. Tests that fail are not ignored. Do you want me to ignore all tests?

Why not the other way round…?

**Comment by goral09** [comment]
> > Tests that pass with stubbed validate_lnc are ignored. Tests that fail are not ignored. Do you want me to ignore all tests?
> 
> Why not the other way round…?

Then the passing tests would be the ones that are passing artificially, b/c they expect what stub returns. That was actually my first proposal but my impression was that you didn't like it and convinced me to do it the other way around :)

File: node/src/components/consensus/highway_core/state.rs (hunk 1 above)
Comment (goral09) [reply node/src/components/consensus/highway_core/state.rs]: Done.

**Comment by afck** [comment]
I'm confused… :sweat_smile: 
I think it's fine to merge the tests that are already passing as they are, and to `#[ignore]` the ones that are not yet passing because `validate_lnc` still returns `None`. The fact that the former will _still_ pass once it's implemented isn't a problem, is it?

**Comment by goral09** [comment]
Yes, I thought so to. I then misunderstood you and convinced myself to do it the other way around ;) That we do not want to have tests passing only b/c they use stubbed values. In theory, we would not catch if they were not using correct code at all b/c they use stubs...

**Pull request merged**